File: ngekos_in_flutter/lib/models/properti.dart
import '../config/api_config.dart';
import '../utils/json.dart';
import 'kamar.dart';

class Properti {
  final int id;
  final String nama;
  final String kota;
  final double? latitude;
  final double? longitude;
  final String alamat;
  final String? deskripsi;
  final List<String>? fasilitas;
  final String? aturan;
  final int? dendaPerHari;
  final int? harga;
  final String? jenisHarga;
  final String status;
  final String? foto;
  final int totalKamar;
  final int kamarTersedia;
  final PemilikInfo? pemilik;
  final List<Kamar>? kamars;

  Properti({
    required this.id,
    required this.nama,
    required this.kota,
    this.latitude,
    this.longitude,
    required this.alamat,
    this.deskripsi,
    this.fasilitas,
    this.aturan,
    this.dendaPerHari,
    this.harga,
    this.jenisHarga,
    required this.status,
    this.foto,
    required this.totalKamar,
    required this.kamarTersedia,
    this.pemilik,
    this.kamars,
  });

  factory Properti.fromJson(Map<String, dynamic> json) {
    final totalKamar = jsonInt(json['total_kamar'], 0);
    final kamarTerisi = jsonInt(json['kamar_terisi'], 0);
    final kamarTersedia =
        jsonInt(json['kamar_tersedia'], totalKamar - kamarTerisi);
    return Properti(
      id: jsonInt(json['id'], 0),
      nama: json['nama'] ?? '',
      kota: json['kota'] ?? '',
      latitude: _doubleOrNull(json['latitude']),
      longitude: _doubleOrNull(json['longitude']),
      alamat: json['alamat'] ?? '',
      deskripsi: json['deskripsi'],
      fasilitas: json['fasilitas'] != null
          ? List<String>.from(json['fasilitas'] as List)
          : null,
      aturan: json['aturan'],
      dendaPerHari: jsonIntOrNull(json['denda_per_hari']),
      harga: jsonIntOrNull(json['harga']),
      jenisHarga: json['jenis_harga'],
      status: json['status'] ?? '',
      foto: ApiConfig.resolveStorageUrl(json['foto']),
      totalKamar: totalKamar,
      kamarTersedia: kamarTersedia,
      pemilik: json['pemilik'] != null
          ? PemilikInfo.fromJson(json['pemilik'] as Map<String, dynamic>)
          : null,
      kamars: json['kamars'] != null
          ? (json['kamars'] as List).map((k) => Kamar.fromJson(k as Map<String, dynamic>)).toList()
          : null,
    );
  }
}

class PemilikInfo {
  final int id;
  final String nama;
  final String? noHp;

  PemilikInfo({required this.id, required this.nama, this.noHp});

  factory PemilikInfo.fromJson(Map<String, dynamic> json) {
    return PemilikInfo(
      id: jsonInt(json['id'], 0),
      nama: json['nama'] ?? '',
      noHp: json['no_hp'],
    );
  }
}

double? _doubleOrNull(dynamic value) {
  if (value == null) return null;
  return double.parse(value.toString());
}

File: app/Models/Properti.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Properti extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'pemilik_id', 'nama', 'kota', 'alamat', 'latitude', 'longitude', 'deskripsi', 'fasilitas', 'aturan', 'denda_per_hari',
        'harga', 'harga_mingguan', 'harga_harian', 'jenis_harga', 'harga_asli', 'status', 'foto',
    ];

    protected function casts(): array
    {
        return [
            'denda_per_hari' => 'decimal:2',
            'harga' => 'decimal:2',
            'harga_mingguan' => 'decimal:2',
            'harga_harian' => 'decimal:2',
            'harga_asli' => 'decimal:2',
            'latitude' => 'decimal:7',
            'longitude' => 'decimal:7',
        ];
    }

    protected static function booted(): void
    {
        $lupakanCache = function (): void {
            foreach (['katalog.markers', 'katalog.daftarKota', 'beranda.markers', 'beranda.daftarKota', 'beranda.kotaStatistik', 'beranda.totalProperti', 'beranda.totalKamar'] as $kunci) {
                try {
                    cache()->forget($kunci);
                } catch (\Throwable $e) {
                }
            }
        };

        static::saved($lupakanCache);
        static::deleted($lupakanCache);
    }

    public function pemilik(): BelongsTo
    {
        return $this->belongsTo(User::class, 'pemilik_id');
    }

    public function admins(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'admin_properti', 'properti_id', 'admin_id');
    }

    public function kamars(): HasMany
    {
        return $this->hasMany(Kamar::class);
    }

    public function fotos(): HasMany
    {
        return $this->hasMany(PropertiFoto::class)->orderBy('urutan')->orderBy('id');
    }

    private ?array $galeriCache = null;

    /**
     * Daftar URL galeri (cover dulu). Fallback ke kolom foto lama bila galeri kosong.
     * Hasil di-memoize per instance agar pemanggilan berulang di satu render gratis.
     *
     * @return array<int,string>
     */
    public function galeriUrls(): array
    {
        if ($this->galeriCache !== null) {
            return $this->galeriCache;
        }

        $dariGaleri = $this->relationLoaded('fotos')
            ? $this->fotos->sortBy([['urutan', 'asc'], ['id', 'asc']])->pluck('path')->all()
            : $this->fotos()->orderBy('urutan')->orderBy('id')->pluck('path')->all();

        $urls = collect($dariGaleri)
            ->filter()
            ->map(fn ($p) => '/storage/'.$p)
            ->values()
            ->all();

        if ($urls === [] && $this->foto) {
            $urls[] = '/storage/'.$this->foto;
        }

        return $this->galeriCache = $urls;
    }

    public function fotoCover(): ?string
    {
        $galeri = $this->galeriUrls();

        return $galeri[0] ?? null;
    }

    /**
     * Harga acuan properti untuk satu periode.
     */
    public function hargaUntuk(string $periode): ?float
    {
        $nilai = match ($periode) {
            'mingguan' => $this->harga_mingguan,
            'harian' => $this->harga_harian,
            default => $this->harga,
        };

        return $nilai !== null ? (float) $nilai : null;
    }

    /**
     * @return array<int,string>
     */
    public function periodeTersedia(): array
    {
        $daftar = [];

        foreach (['bulanan', 'mingguan', 'harian'] as $periode) {
            $harga = $this->hargaUntuk($periode);

            if ($harga !== null && $harga > 0) {
                $daftar[] = $periode;
            }
        }

        return $daftar;
    }

    public function peminat(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'properti_favorits', 'properti_id', 'user_id')->withTimestamps();
    }

    public function ulasans(): HasMany
    {
        return $this->hasMany(Ulasan::class);
    }
}
